sway-lsp/src/capabilities/semantic_tokens.rs

```diff
@@ -28,43 +34,51 @@ pub fn get_semantic_tokens_full(
     }
 }
 
-pub fn to_semantic_tokes(tokens: &[Token]) -> Vec<SemanticToken> {
-    if tokens.is_empty() {
-        return vec![];
-    }
+pub fn to_semantic_tokens(token_map: &TokenMap) -> Vec<SemanticToken> {
+    let mut semantic_tokens: Vec<SemanticToken> = Vec::new();
 
-    let mut semantic_tokens: Vec<SemanticToken> = vec![create_semantic_token(&tokens[0], None)];
+    let mut prev_token_span = None;
+    for ((_, span), token) in token_map.iter() {
+        let token_type_idx = get_type_idx(&token.parsed);
+        let semantic_token = create_semantic_token(token_type_idx, span, prev_token_span);
 
-    for i in 1..tokens.len() {
-        let semantic_token = create_semantic_token(&tokens[i], Some(&tokens[i - 1]));
         semantic_tokens.push(semantic_token);
+        prev_token_span = Some(span);
     }
 
     semantic_tokens
 }
 
-fn create_semantic_token(next_token: &Token, prev_token: Option<&Token>) -> SemanticToken {
+fn create_semantic_token(
+    token_type_idx: u32,
+    next_token_span: &Span,
+    prev_token_span: Option<&Span>,
+) -> SemanticToken {
+    let next_token_range = get_range_from_span(next_token_span);
+    let next_token_line_start = next_token_range.start.line;
+
     // TODO - improve with modifiers
     let token_modifiers_bitset = 0;
-    let token_type = get_type(&next_token.token_type);
-    let length = next_token.length;
+    let length = next_token_range.end.character - next_token_range.start.character;
```

Comment:
```suggestion
    let length = next_token_range.len_chars();
```
^ Maybe this range type should have a `len_chars()` method or something along these lines?

Reply:
Yeah that would be cool. The `next_token_range` here is actually a `Range` type in the `lsp_types` crate though.